### PR TITLE
VideoPress: Validate embedded post context when resolving playback access

### PR DESCRIPTION
**Pull request by obenland**
Fixes #

## Proposed changes
* In `Access_Control::is_current_user_authed_for_video()`, validate the supplied `embedded_post_id` against the requested `guid` before using it as authorization context.
* When the post id cannot be proven to embed the requested guid (via the `videopress/video` block's `guid` attribute or a `[videopress]`/`[wpvideo]` shortcode whose positional argument matches), normalize it to `0`. The existing `build_restriction_details()` flow already returns `can_access: false` for `post_id = 0`, so no other logic needs to change.
* Public videos, the `current_user_can( 'upload_files' )` short-circuit, and the existing memberships/paywall paths are unchanged.
* Add a dedicated test file `Video_Authorization_Test.php` covering the affected paths.

## Related product discussion/links
* 

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions
* In `projects/packages/videopress/`, run `composer phpunit` — all tests, including the new `Video_Authorization_Test`, pass (106/106).
* Run `composer phpcs:lint` from the repo root on the modified files (`projects/packages/videopress/src/class-access-control.php` and `projects/packages/videopress/tests/php/Video_Authorization_Test.php`) — no new violations.
* On a local Jetpack site with VideoPress connected:
  * Upload a private VideoPress video and embed it on a draft/publish post via the `videopress/video` block. As a subscriber with read access to the post, confirm playback still works (browser console shows a successful JWT response from `/wp-admin/admin-ajax.php?action=videopress-get-playback-jwt`).
  * Confirm a public VideoPress video continues to play for both subscribers and logged-out visitors.
  * Confirm a contributor/author with `upload_files` continues to receive a playback JWT for any video.

### CODE REVIEW

File: projects/packages/videopress/src/class-access-control.php [hunk 1]
```diff
@@ -255,6 +255,107 @@ private function default_video_restriction_details( $default_can_access = false
...
+	 * @param string $post_content The post content to scan.
+	 * @param string $guid         The video guid to match.
+	 *
+	 * @return bool
+	 */
+	private function post_content_has_videopress_block( $post_content, $guid ) {
+		if ( false === strpos( $post_content, 'wp:videopress/video' ) ) {
+			return false;
+		}
+
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/class-access-control.php:307]: Applied — added a single `strpos( $post->post_content, $guid )` check at the top of `post_embeds_videopress_guid()` so the optimization benefits the shortcode scan path too. Pushed in 3fce231.

File: projects/packages/videopress/src/class-access-control.php [hunk 2]
```diff
@@ -288,6 +389,11 @@ public function is_current_user_authed_for_video( $guid, $embedded_post_id, $sel
 			return false;
 		}
 
+		$embedded_post_id = (int) $embedded_post_id;
+		if ( $embedded_post_id && ! $this->post_embeds_videopress_guid( $embedded_post_id, $guid ) ) {
+			$embedded_post_id = 0;
+		}
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/class-access-control.php:437]: Applied — gated the validator on `VIDEOPRESS_PRIVACY::IS_PUBLIC !== $video_info->privacy_setting` so public videos take the fast path while private and site-default videos still validate. Pushed in 773e15c. (Did not move the validator into the switch branches because the SITE_DEFAULT case may also resolve to private and would need the validator there too — gating in the existing if is simpler.)

File: projects/packages/videopress/src/class-access-control.php [hunk 3]
```diff
@@ -255,6 +255,112 @@ private function default_video_restriction_details( $default_can_access = false
...
+		if ( false === strpos( $post->post_content, $guid ) ) {
+			return false;
+		}
+
+		if ( $this->post_content_has_videopress_block( $post->post_content, $guid ) ) {
+			return true;
+		}
+
+		return $this->post_content_has_videopress_shortcode( $post->post_content, $guid );
+	}
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/class-access-control.php:292]: Applied — added `post_content_has_videopress_url()` covering the canonical VideoPress hosts (`videos.videopress.com`, `videos.files.wordpress.com`, `videopress.com/v|embed`, `v.wordpress.com`). That single URL check accepts oEmbed URLs, `core/embed`, `core/video`, and core `[video]` shortcodes that resolve to a VideoPress URL. Pushed in 94ae9da.

File: projects/packages/videopress/tests/php/Video_Authorization_Test.php [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+				'post_content' => 'Intro [videopress ' . $guid . '] trailing.',
+				'post_status'  => 'publish',
+			)
+		);
+		$this->set_current_user_role( 'subscriber' );
+
+		$this->assertTrue( Access_Control::instance()->is_current_user_authed_for_video( $guid, $embedding ) );
+	}
+
+	/**
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/tests/php/Video_Authorization_Test.php:116]: Partially applied — added `test_subscriber_with_videopress_url_embed_post_id_is_authorized_for_private_video` covering the URL-based embed path (which also exercises `core/embed` semantically since the URL inside the figure is what the validator matches). Did not add the pure `<!-- wp:videopress/video {\"guid\":...} /-->` block test: WorDBless's SQLite polyfill stores block JSON with escaped quotes (`{\\"guid\\":...`), which makes `parse_blocks()` return `attrs => null` and the test would fail for an environmental reason rather than a code reason. The block-parsing branch is exercised in production but tested indirectly here by the negative cases that prove the validator returns different answers based on guid content. Pushed in 94ae9da.

File: projects/packages/videopress/src/utility-functions.php [hunk 5]
```diff
@@ -34,6 +34,34 @@ function videopress_is_valid_guid( $guid ) {
...
+function videopress_extract_guid_from_url( $url ) {
+	if ( ! is_string( $url ) || '' === $url ) {
+		return null;
+	}
+
+	// phpcs:ignore WordPress.WP.CapitalPDangit.MisspelledInText -- regex matches literal hostnames.
+	$patterns = array(
+		'@videos\.(?:videopress\.com|files\.wordpress\.com)/([a-z0-9]{8})/@i',
+		'@(?:videopress\.com/(?:v|embed)|video\.wordpress\.com/v)/([a-z0-9]{8})@i',
+		'|^https?://v\.wordpress\.com/([a-z0-9]{8})(?:[/?#].*)?$|i',
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/utility-functions.php]: Applied — reused `Utils::is_videopress_url()`'s canonical regex by adding `Utils::extract_videopress_guid_from_url()` that shares the same pattern and refactoring `is_videopress_url()` to delegate to it. The pattern already covers `video.wordpress.com/embed/<guid>` and the unconstrained `v.wordpress.com/<guid>` form. Pushed in 133faf1.

File: projects/packages/videopress/src/utility-functions.php [hunk 6]
```diff
@@ -34,6 +34,34 @@ function videopress_is_valid_guid( $guid ) {
...
+	);
+
+	foreach ( $patterns as $pattern ) {
+		if ( preg_match( $pattern, $url, $matches ) ) {
+			return $matches[1];
+		}
+	}
+
+	return null;
+}
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/utility-functions.php]: Applied — added `test_extract_videopress_guid_from_url_returns_guid_for_supported_forms` and `test_extract_videopress_guid_from_url_returns_null_for_unsupported_urls` in `UtilsTest.php` covering every URL shape `Utils::is_videopress_url()` accepts (including `video.wordpress.com/embed/` and `v.wordpress.com/`). Pushed in 133faf1.

File: projects/packages/videopress/src/utility-functions.php [hunk 7]
```diff
@@ -34,6 +34,34 @@ function videopress_is_valid_guid( $guid ) {
...
+	if ( ! is_string( $url ) || '' === $url ) {
+		return null;
+	}
+
+	// phpcs:ignore WordPress.WP.CapitalPDangit.MisspelledInText -- regex matches literal hostnames.
+	$patterns = array(
+		'@videos\.(?:videopress\.com|files\.wordpress\.com)/([a-z0-9]{8})/@i',
+		'@(?:videopress\.com/(?:v|embed)|video\.wordpress\.com/v)/([a-z0-9]{8})@i',
+		'|^https?://v\.wordpress\.com/([a-z0-9]{8})(?:[/?#].*)?$|i',
+	);
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/utility-functions.php]: Applied for the URL-to-guid path — `Utils::extract_videopress_guid_from_url()` is now the single source of truth for that regex and `is_videopress_url()` delegates to it. `Block_Editor_Content::video_shortcode_override()` also recognises the `videos.videopress.com/<guid>/` and `videos.files.wordpress.com/<guid>/` file-host variants that aren't shared by the oEmbed provider set — centralising those is a legitimate follow-up but outside this PR's scope since it would touch an unrelated call path. Pushed in 133faf1.

File: projects/packages/videopress/src/class-utils.php [hunk 8]
```diff
@@ -74,10 +74,23 @@ public static function get_video_press_url( $guid, $attributes = array() ) {
...
+	/**
+	 * Extract the VideoPress guid from a recognised VideoPress URL.
+	 *
+	 * @param string $url The URL to inspect.
+	 * @return string|null The 8-character guid, or null if the URL is not a recognised VideoPress URL.
+	 */
+	public static function extract_videopress_guid_from_url( $url ) {
 		if ( ! is_string( $url ) ) {
-			return false;
+			return null;
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/class-utils.php:88]: Applied — changed the phpdoc to `@param mixed $url` to match the runtime contract (non-strings are accepted and return null). Pushed in 1ef3a3d.

File: projects/packages/videopress/tests/php/Video_Authorization_Test.php [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+				'post_content' => '<figure><div>https://videopress.com/v/' . $guid . '</div></figure>',
+				'post_status'  => 'publish',
+			)
+		);
+		$this->set_current_user_role( 'subscriber' );
+
+		$this->assertTrue( Access_Control::instance()->is_current_user_authed_for_video( $guid, $embedding ) );
+	}
+
+	/**
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/tests/php/Video_Authorization_Test.php:98]: Applied — added `test_subscriber_with_file_host_videopress_url_embed_post_id_is_authorized_for_private_video` using a `videos.videopress.com/<guid>/original.mp4` URL. Also extended `Utils::extract_videopress_guid_from_url()` so the file-host forms are now a recognised VideoPress URL shape across the package. Pushed in 1ef3a3d.

File: projects/packages/videopress/src/class-access-control.php [hunk 10]
```diff
@@ -255,6 +255,145 @@ private function default_video_restriction_details( $default_can_access = false
...
+	 */
+	private function post_content_has_videopress_url( $post_content, $guid ) {
+		if ( ! preg_match_all( '#https?://[^\s"\'<>)]+#i', $post_content, $matches ) ) {
+			return false;
+		}
+
+		foreach ( $matches[0] as $url ) {
+			if ( $guid === Utils::extract_videopress_guid_from_url( $url ) ) {
+				return true;
+			}
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/class-access-control.php:391]: Applied — extended `Utils::extract_videopress_guid_from_url()`'s regex to include `videos.videopress.com/<guid>/` and `videos.files.wordpress.com/<guid>/`, so the validator's URL scan now accepts the file-host forms without any local regex in `post_content_has_videopress_url()`. Covered by new UtilsTest cases and a dedicated playback authorization test. Pushed in 1ef3a3d.

File: projects/packages/videopress/src/class-utils.php [hunk 11]
```diff
@@ -74,10 +74,23 @@ public static function get_video_press_url( $guid, $attributes = array() ) {
 	 * @return bool True if the URL is a VideoPress URL, false otherwise.
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/class-utils.php]: Applied — updated `is_videopress_url()`'s phpdoc to `@param mixed $url` to mirror the extractor it delegates to. Pushed in 36beb5c.

File: projects/packages/videopress/src/class-access-control.php [hunk 12]
```diff
@@ -288,6 +427,15 @@ public function is_current_user_authed_for_video( $guid, $embedded_post_id, $sel
...
 		}
 
+		$embedded_post_id = (int) $embedded_post_id;
+		if (
+			$embedded_post_id
+			&& VIDEOPRESS_PRIVACY::IS_PUBLIC !== $video_info->privacy_setting
+			&& ! $this->post_embeds_videopress_guid( $embedded_post_id, $guid )
+		) {
+			$embedded_post_id = 0;
+		}
```
[the remark this answers is not in the document]
Comment (obenland) [reply projects/packages/videopress/src/class-access-control.php:437]: Not applied — `Data::get_videopress_videos_private_for_site()` reads a WordPress option that's served from the in-memory options cache on repeated calls, so calling it a second time in the gate adds no meaningful overhead. The extra variable and condition make the code harder to follow without a real benefit.

**Review by jeherve — APPROVED** [review]
This looks good to me, and should be good to merge! 🚢
